webapp/src/main/java/com/box/l10n/mojito/common/notification/IntegrityCheckNotifier.java

```diff
@@ -0,0 +1,99 @@
+import jakarta.annotation.PostConstruct;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+
+/**
+ * Sends warning notifications to the specified warning channel on Slack. Selects default Slack
+ * client specified in app properties. The channel name for warnings must be placed in the app
+ * properties under: l10n.integrity-check-notifier.slackChannel
+ *
+ * @author mattwilshire
+ */
+@Component
+@ConditionalOnProperty(value = "l10n.integrity-check-notifier.enabled", havingValue = "true")
+public class IntegrityCheckNotifier {
+
+  static Logger logger = LoggerFactory.getLogger(IntegrityCheckNotifier.class);
+
+  private final SlackClients slackClients;
+  private final TMTextUnitRepository tmTextUnitRepository;
+  private final IntegrityCheckNotifierConfiguration integrityCheckNotifierConfiguration;
+  private final SlackMessageBuilder slackMessageBuilder;
+  private SlackClient slackClient;
+
+  @Autowired
+  public IntegrityCheckNotifier(
+      SlackClients slackClients,
+      TMTextUnitRepository tmTextUnitRepository,
+      IntegrityCheckNotifierConfiguration integrityCheckNotifierConfiguration,
+      SlackMessageBuilder slackMessageBuilder) {
+    this.slackClients = slackClients;
+    this.tmTextUnitRepository = tmTextUnitRepository;
+    this.integrityCheckNotifierConfiguration = integrityCheckNotifierConfiguration;
+    this.slackMessageBuilder = slackMessageBuilder;
+  }
+
+  @PostConstruct
+  public void init() throws IntegrityCheckNotifierException {
+    // This method will only be called if slack-warnings are enabled in app properties.
+    String slackClientId = integrityCheckNotifierConfiguration.getSlackClientId();
+
+    if (slackClientId == null)
+      throw new IntegrityCheckNotifierException("Slack client id not defined.");
+
+    if (integrityCheckNotifierConfiguration.getSlackChannel() == null)
+      throw new IntegrityCheckNotifierException("Slack channel not defined.");
+
+    if (!integrityCheckNotifierConfiguration.getSlackChannel().startsWith("#"))
+      throw new IntegrityCheckNotifierException("Slack channel must start with #.");
+
+    this.slackClient = slackClients.getById(slackClientId);
+
+    if (slackClient == null) {
+      throw new IntegrityCheckNotifierException("Slack client id defined but doesn't exist.");
+    }
+  }
+
+  public void handleIntegrityException(
+      IntegrityCheckException exception, Long tmTextUnitId, String targetString) {
+    if (slackMessageBuilder == null) return;
+
+    // Lookup text unit for repository and source text.
+    TMTextUnit tmTextUnit = tmTextUnitRepository.findById(tmTextUnitId).orElse(null);
+    try {
+      // Branch off this if statement to send custom warning messages.
+      if (exception instanceof HtmlTagIntegrityCheckerException) {
```

Comment:
Can we move this into the integrity checker itself, so that the integrity checker has a reference to the notifier and it triggers the notification logic prior to raising the exception?